packages/controls/src/controls/shape-v2/shape-v2.ts

```diff
@@ -0,0 +1,322 @@
+  } as const
+  static readonly type = 'makeswift::controls::shape-v2' as const
+
+  static readonly Layout = {
+    Inline: `${this.type}::layout::inline`,
+    Popover: `${this.type}::layout::popover`,
+  } as const
+
+  static deserialize(
+    data: DeserializedRecord,
+    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
+  ): ShapeV2Definition {
+    if (data.type !== Definition.type)
+      throw new Error(
+        `Shape: expected '${Definition.type}', got '${data.type}'`,
+      )
+
+    const {
+      config: { type, ...config },
+    } = z
+      .object({
+        type: z.literal(Definition.type),
+        config: z.object({
+          label: z.string().optional(),
+          layout: z
+            .union([
+              z.literal(Definition.Layout.Inline),
+              z.literal(Definition.Layout.Popover),
+            ])
+            .optional(),
+          type: z.record(z.string(), SerializationSchema.deserializedRecord),
+        }),
+      })
+      .parse(data)
+
+    const deserializedType = mapValues(type, (itemDef) => {
+      return deserializeCallback(itemDef)
+    })
+
+    return ShapeV2({
+      type: deserializedType,
+      ...config,
+    })
+  }
+
+  get keyDefs() {
+    return this.config.type
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    const version = z.literal(1)
+    const type = z.literal(Definition.type)
+
+    const keys = mapValues(this.keyDefs, (def) => def.schema)
+
+    const v0Data = z.object(mapValues(keys, (key) => key.data)) as SchemaType<
+      ShapeV2DataTypeV0<C>
+    >
+    const v1Data = z.object({
+      [ControlDataTypeKey]: z.literal(Definition.v1DataType),
+      value: v0Data,
+    }) as SchemaType<ShapeV2DataTypeV1<C>>
+
+    const data = z.union([v0Data, v1Data])
+
+    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
+      ValueType<C>
+    >
+
+    const resolvedValue = z.object(
+      mapValues(keys, (key) => key.resolvedValue),
+    ) as SchemaType<ResolvedValueType<C>>
+
+    const config = z.object({
+      label: z.string().optional(),
+      layout: z.union([
+        z.literal(Definition.Layout.Inline),
+        z.literal(Definition.Layout.Popover),
+      ]),
+      type: z.object(mapValues(keys, (key) => key.definition)),
+    })
+
+    const definition = z.object({
+      type,
+      config,
+    })
+
+    return { type, data, value, resolvedValue, definition, version }
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    if (data == null) return undefined
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.fromData((data as ShapeV2DataTypeV1<C>).value[key]),
+      )
```

Comment:
I _think_ we should be able to avoid casting here if we use `ts-pattern` along the lines of https://github.com/makeswift/makeswift/blob/main/packages/controls/src/controls/image/image.ts#L202-L203

Reply:
Let me know if you want to work through this together.

Reply:
Actually, since the only difference between v1 and v2 data is whether the value is nested inside of the `value` prop, ideally we shouldn't have two branches of the code that processes it and have something like this instead:

```typescript
const shapeData = toShapeData(data)
return mapValues(this.keyDefs, (def, key) =>
        def.fromData(data[key]),
      )
```

Reply:
I tried `ts-pattern` and it wasn't working I think because of v1 is a sub type of v0. 

I **love** the idea of `toShapeData` so I'll try that and also do something similar in leo

Reply:
This the error I get from doing something similar with `ts-pattern`

![CleanShot 2024-12-20 at 10 19 29@2x](https://github.com/user-attachments/assets/61c2e5c7-f931-411a-b971-9b2ca0c5f6bf)

![CleanShot 2024-12-20 at 10 19 23@2x](https://github.com/user-attachments/assets/bb6154f7-6c9e-4d5c-9a6f-50552de0e074)

Reply:
Ahh geez. Looks like there was an error in my implementation that could have caused this error. looking...

Reply:
Just kidding. I think we should update the data instead of downgrading it. Adding a. method to the definition
```

  versionData(data: DataType<C>): ShapeV2DataTypeV1<C> {
    if (data == null || !(ControlDataTypeKey in data)) {
      return {
        [ControlDataTypeKey]: Definition.v1DataType,
        value: data,
      }
    }
    return data as ShapeV2DataTypeV1<C>
  }
```

This is kinda like the `normalizeData` in the `RichTextV2` control